File: pedigree/ped_draw.py
''' MODULE: ped_draw.py

This is a pedigree plotter implemented with Graphviz. It needs some serious
work.

The output first goes to an intermediate file *.dot and then to the final *.pdf
It currently cannot handle multiple generations and has a slight bug for
exactly 2 or 3 children. It looks great for a single marriage with 6 children!

See file ped_draw_example.py for an example of how to use this module
'''

from graphviz import Graph
import ped_solve as ps

class Draw():
    def __init__(self, name, format='pdf'):
        self.name = name
        self.next_hidden = 0  # next available hidden node ID
        self.nodes = []  # list of currently existing named nodes

        # create the graph and set attributes for pedigree
        self.dot = Graph(comment=name, format=format)
        self.dot.graph_attr['splines'] = 'ortho'
        self.dot.node_attr['fontname'] = 'helvetica'
        self.dot.node_attr['fixedsize'] = 'true'
        self.dot.node_attr['style'] = 'filled'
        self.dot.node_attr['height'] = '0.5'
        self.dot.body.append('\tlabelloc="t";')
        self.dot.body.append('\tlabel="'+name+'";')


    #########################
    # BASIC DRAWING FUNCTIONS
    #########################

    def hidden(self, names):
        # define hidden nodes from a list
        for name in names:
            self.dot.node(name, width='0', shape='point', style='invis')

    def indiv(self, person):
        # define an individual node from a Person
        shape = 'square'
        if person.sex == ps.Sex.f:
            shape = 'circle'

        # shade the node based on infection probability
        R = 255
        G = int(119 + 136*(1-person.infected))
        B = int(86 + 169*(1-person.infected))
        color = '#'+hex(R)[2:]+hex(G)[2:]+hex(B)[2:]

        self.dot.node(person.name, shape=shape, fillcolor=color)

    def srank(self, nodes):
        # places a list of nodes in the same rank
        self.dot.body.append(
            '\t{rank=same; ' +
            ' -- '.join([n for n in nodes]) +
            ';}'
        )

    def edges(self, edgelist, constraint='true'):
        # makes a bunch of edges in 2D list
        for pair in edgelist:
            self.dot.edge(pair[0], pair[1], constraint=constraint)

    def render(self):
        # render self to DOT, but do not display
        return self.dot.source

    def show(self):
        # render and display self
        # return generated DOT source for convenience
        self.dot.render(self.name+'.dot', view=True)
        return self.dot.source


    ###########################
    # FAMILY PLOTTING FUNCTIONS
    ###########################

    def draw_marriage(self, father, mother, children, silent=False):
        # accepts two Person objects and draws the marriage between them
        # takes a list of children as arg until i figure out how to do that
        # silent allows no stdout printing for CGI script

        # if unknown, find the probability
        # reject the probability if invalid
        for p in [father, mother]+children:
            if ps.find_prob(p) < 0:
                if not silent:
                    print 'err: unable to determine probability for individual %s'\
                       % p.get()[0]
                return -1  # failure

        # append these people to our list of nodes
        if father.name not in self.nodes:
            self.indiv(father)
            self.nodes += [father.name]
        if mother.name not in self.nodes:
            self.indiv(mother)
            self.nodes += [mother.name]
        self.nodes += [c.name for c in children]

        if len(children) == 0:
            # no children = no tricky stuff with hidden nodes!
            self.srank([father.name, mother.name])
        else:
            # make a hidden node in the middle to connect to children
            self.hidden([str(self.next_hidden)])
            self.srank([father.name, str(self.next_hidden), mother.name])
            self.next_hidden += 1

        if len(children) == 1:
            # there is only one child, so no branching for children
            self.indiv(children[0])
            self.nodes += children[0].name
            self.edges([[str(self.next_hidden-1), children[0].name]])
        elif len(children) > 1:
            # general branching required for children
            N = len(children) - (len(children)%2) - 1  # hidden nodes needed
            hnodes = [str(h) for h in range(self.next_hidden,
                                            self.next_hidden+N)]
            self.hidden(hnodes)
            self.srank(hnodes)

            # link center node to marriage node
            self.edges([[str(self.next_hidden-1), str(self.next_hidden+N/2)]])
            for c in children:
                self.indiv(c)

            # link the first child to the correct node
            self.edges([[str(self.next_hidden), children[0].name]])

            # link subsequent children to the corresponding nodes
            edgelist = []
            for i, c in enumerate(children[1:-1]):
                add = 0
                if len(children)%2 == 0 and i >= N/2:
                    add = 1
                edgelist += [[str(self.next_hidden+i+add), c.name]]
            self.edges(edgelist)

            # link the final child to the correct node
            self.edges([[str(self.next_hidden+N-1), children[-1].name]])
            self.next_hidden += N

            return 1  # success

File: js/ped_draw.js
/* =============
File: ped_draw.js
Authors: Eric, Nelson, and Karena
Course: CSE
Description: module containing functions for plotting a pedigree
  assumes that ped_solve.js has already been included with a <script> tag
============= */

const ped_draw = (function () {
  return {
    Draw: function (title) {
      this.src = "";  // String of Dot source code to render (validate!)
      this.next_hidden = 0;  // next available hidden node ID
      this.nodes = [];  // list of currently existing named nodes

      // create graph and set pedigree attributes
      this.src += `graph {
        graph[splines=ortho];
        node [fixedsize=true fontname=helvetica height=0.5 style=filled];
        labelloc="t";
        label="${title}";`;


      /* BASIC DRAWING FUNCTIONS */

      this.hidden = names => {
        // define hidden nodes from a list
        names.forEach(name =>
          this.src += `"${name}" [shape=point style=invis width=0];`);
      }

      this.indiv = person => {
        // define an individual node from a ped_solve.Person object
        let shape = "square";
        if (person.sex === ped_solve.Sex.f)
          shape = "circle";
        var penetrance = 0;
        if (person.infected === true){
          penetrance = 1;
        }
        // shade the node based on infection probability
        let RGB = [255,
                   Math.round(119 + 136 * (1-penetrance)),
                   Math.round(86 + 169 * (1-penetrance))];
        let color_as_hex = "#"+RGB.map(d =>
          ("0"+d.toString(16)).slice(-2)).join("");

        this.src += `"${person.name}"
          [fillcolor="${color_as_hex}" shape=${shape}];`;
      }

      this.srank = nodes => {
        // places a list of nodes in the same rank
        this.src += `{rank=same; "${nodes.join('" -- "')}";};`;
      }

      this.edges = (edgelist, constraint="true") => {
        // makes a bunch of edges in 2D list
        edgelist.forEach(pair =>
          this.src += `"${pair[0]}" -- "${pair[1]}"
            [constraint=${constraint}];`);
      }

      this.render = () => {
        // return self as DOT language string
        return this.src + "}";
      }


      /* FAMILY PLOTTING FUNCTIONS */

      this.draw_marriage = (father, mother, children) => {
        // accepts two Person objects and draws the marriage between them
        // takes a list of children as arg until i figure out how to do that
        // we'll just assume that the pedigree is valid (error check later?)

        // if unknown, find the probability of disease
        //[father, mother, ...children].forEach(ped_solve.find_prob);

        // append the new people to our list of nodes
        [father, mother]
          .filter(person => !this.nodes.includes(person.name))
          .forEach(person => {
            this.indiv(person);
            this.nodes.push(person.name);
          });
        this.nodes = this.nodes.concat(children.map(child => child.name));

        if (children.length === 0) {
          // no children = no tricky stuff with hidden nodes!
          this.srank([father.name, mother.name]);
        } else {
          // make a hidden node in the middle to connect to children
          this.hidden([this.next_hidden.toString()]);
          this.srank([father.name, this.next_hidden.toString(), mother.name]);
          this.next_hidden++;
        }

        if (children.length === 1) {
          // there is only one child, so no branching for children
          this.indiv(children[0]);
          this.nodes.push(children[0].name);
          this.edges([[(this.next_hidden - 1).toString(), children[0].name]]);
        } else if (children.length > 1){
          // general branching required for children

          // first count number of hidden nodes needed
          let N = Math.round(children.length - (children.length % 2) - 1);
          let hnodes = [...Array(N).keys()]
                        .map(x => x + this.next_hidden)
                        .map(x => x.toString());
          this.hidden(hnodes);
          this.srank(hnodes);

          // link center node to marriage node
          this.edges([[(this.next_hidden - 1).toString(),
                       (this.next_hidden + Math.trunc(N / 2)).toString()]]);
          children.forEach(this.indiv);

          // link the first child to the correct node
          this.edges([[this.next_hidden.toString(), children[0].name]]);

          // link subsequent children to corresponding nodes
          edgelist = children.slice(1, -1).map((child, i) => {
            let add = 0;
            if (children.length % 2 === 0 && i >= Math.trunc(N / 2))
              add = 1;
            return [(this.next_hidden + i + add).toString(), child.name];
          });
          this.edges(edgelist);

          // link the final child to the correct node
          this.edges([[(this.next_hidden + N - 1).toString(),
                       children.slice(-1)[0].name]]);
          this.next_hidden += N;
        }
      }
    }
  }
}());
